src/QuadForm/Spaces.jl

```diff
@@ -273,6 +273,8 @@ inner_product(V::AbstractSpace, v::MatElem, w::MatElem)
 
 _inner_product(L::AbstractLat, v, w) = inner_product(ambient_space(L), v, w)
 
+inner_product(L::AbstractLat, v, w) = inner_product(ambient_space(L), v, w)
```

Comment:
This is the opposite of what is added for `ZZLat`, or am I missing something?

Reply:
We have to figure out if $v$ and $w$ are given with respect to the ambient space or are vectors with respect to the basis of $L$.  I would argue that 
`inner_product(L::ZLat, u::ZZMat, v::ZZMat)`
`inner_product(L::ZLat, u::Vector{ZZRingElem}, v::Vector{ZZRingElem})`
indicate that $u$ and $v$ are given w.r.t the basis of $L$. 

`inner_product(L::ZLat, u::QQMat, v::QQMat)`
`inner_product(L::ZLat, u::Vector{QQRingElem}, v::Vector{QQRingElem})`
indicate that $u$ and $v$ are given w.r.t the ambient space of $L$.

Reply:
Hm, I am against it. It don't think that `inner_product(L, [ZZ(1)], [ZZ(1)])` and `inner_product(L, [QQ(1)], [QQ(1)])` should do something different. I would argue that this is not a nice interface for the user and very confusing. At the same time it goes against the (implicit) convention, that promotion should not change the result. It is then also ambiguous what `inner_product(L, [1], [1])` should be.

Reply:
Then I would say that 
`inner_product(V::AbsSpace, u, v)` means take the gram matrix of the space $V$ 
and 
`inner_product(L::ZZLat u, v)` means take the gram matrix of the lattice.... but this does not make sense over number fields ... damn it.

Reply:
I think remember now why I did not add initially. We could allow `inner_product(L, v, w)` only for Z-lattice and integral v, w. This would be better than nothing.

Reply:
I concur with @thofma, this is a no-go.

How about making the form an explicit object, and use it explicitly? And/or introducing a notion of "space with form" ?

Reply:
We do have a space with form.  I decided for the minimally invasive solution that
`inner_product(L::AbstractLat, u v) = inner_product(ambient_space(L), u, v)`